### PR TITLE
systemctl: list-dependencies: support --type= and --state=

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

File: src/systemctl/systemctl-util.c [hunk 1]
```diff
@@ -148,6 +148,33 @@ int get_state_one_unit(sd_bus *bus, const char *unit, UnitActiveState *ret_activ
         return 0;
 }
 
+int get_sub_state_one_unit(sd_bus *bus, const char *unit, char **ret_sub_state) {
+        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_free_ char *sub_state = NULL, *dbus_path = NULL;
+        int r;
+
+        assert(unit);
```
Comment (bluca) [inline src/systemctl/systemctl-util.c:158]: assert(bus)

File: src/basic/strv.h [hunk 2]
```diff
@@ -20,6 +20,16 @@ char* strv_find_startswith(char * const *l, const char *name) _pure_;
 
 #define strv_contains(l, s) (!!strv_find((l), (s)))
 #define strv_contains_case(l, s) (!!strv_find_case((l), (s)))
+#define STRV_CONTAINS_ONE(s, ...)                       \
```
Comment (bluca) [inline src/basic/strv.h]: please add unit tests in src/test/test-strv.c for this

**Comment by bluca** [comment]
Please add some tests for the new options in `test/units/testsuite-26.sh`

**Review by bluca — APPROVED** [review]

File: src/basic/strv.h [hunk 3]
```diff
@@ -20,6 +20,16 @@ char* strv_find_startswith(char * const *l, const char *name) _pure_;
...
+#define STRV_CONTAINS_ONE(s, first, ...)                        \
+        ({                                                      \
+                bool _found = false;                            \
+                FOREACH_STRING(l, first, __VA_ARGS__) {         \
+                        _found = strv_contains(s, l);           \
+                        if (_found)                             \
+                                break;                          \
+                }                                               \
+                _found;                                         \
+        })
```
Comment (poettering) [inline src/basic/strv.h]: this already exists in `strv_overlap()

File: src/systemctl/systemctl-list-dependencies.c [hunk 4]
```diff
@@ -80,6 +80,12 @@ static int list_dependencies_one(
         typesafe_qsort(deps, strv_length(deps), list_dependencies_compare);
 
         STRV_FOREACH(c, deps) {
+                _cleanup_free_ char *load_state = NULL, *sub_state = NULL;
+                UnitActiveState active_state;
+
+                /* We won't be able to preserve the tree structure if --type= or --state= is used */
+                arg_plain = arg_plain || arg_types || arg_states;
```
Comment (poettering) [inline src/systemctl/systemctl-list-dependencies.c]: this should be moved outside of the SRV_FOREACH() loop, it touches global stat eafter al, and is result of global state only

**Pull request merged**